### DISCUSSION TITLE
blog/introducing-x-state-store

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#108] [Announcements]
# blog/introducing-x-state-store

There's a new state manager in the game, and it ticks all my boxes ...

https://tkdodo.eu/blog/introducing-x-state-store#inject-comments

**Comment by kimsagro1** [comment]
Should `useFish` in the first code sample be returning `state.context.fish` rather than `state.context.bears`?

**Reply by TkDodo** [reply]
thanks, already fixed.

**Comment by swalker326** [comment]
Is the store to be exported or is there some different way you're getting access to it?

**Reply by TkDodo** [reply]
yes, you can export it, or you can just stick it into react-context, similar to how you'd do it with zustand: https://tkdodo.eu/blog/zustand-and-react-context

**Reply by Fitzpaa** [reply]
So would we get the same benefits from sticking our xstate store into a react-context that we would get from sticking Zustand into a react-context like you displayed in that blog post?

**Reply by TkDodo** [reply]
yes, that's pretty much the same approach

**Comment by bojanraguz1992** [comment]
How does it fit with NextJS? I believe everything should work as expected, but I'm interested in persistence functionality.
We're currently using Zustand and we had to make some tweaks to persist the store because we had to set some configs within `localStorage` and it was a pain a bit :)

**Reply by TkDodo** [reply]
I don't think there is a persist integration. I've always found persisting on a client storage to be quite hard with SSR in general.

**Reply by TkDodo** [reply]
David answered on [Twitter](https://x.com/DavidKPiano/status/1871176125071962466) that you can just do:

```
store.subscribe(state => {
  persist(state);
});
```

which is pretty great

**Comment by qiushiyan** [comment]
Thanks for the post. Just migrated some zustand stores to xstate store, I personally think the separation between states and transitions help me keep a clearer mental model. 

For persistence, I have been creating the store inside a react context, use `localStorage` to populate the initial state, and use `store.subscribe` to sync localStorage state transitions.

**Reply by qiushiyan** [reply]
Weird things do happen in next.js dev server, but generally this simple persistence pattern works pretty well in production

**Comment by popuguytheparrot** [comment]
```javascript
store.send({ type: 'increasePopulation', by: 10 }) 
```

Redux like api so WORST and WEIRD and CURSED

**Reply by TkDodo** [reply]
I (historically) get that sentiment, but it's easy to forget the upsides of this approach. Let's imagine we just have `store.setPopulation(newPopulation)` and that just sets it into the store. The problem with setters is that we can't easily change logic. It's hard to come up with business logic for counter examples, but let's assume that population cannot exceed `maxPopulation` - another store state. If it would increase `maxPopulation`, it would just be set to the maximum, no matter how much you try to increase. Kinda like a safety valve or something.

Now with event driven architecture, this is easy. You go to the store, and add an if statement to the `increasePopulation ` event.

If you have `store.setPopulation(currentPopulation + 10)` scattered around in your app, you have to go to all buttons / event handlers and add that check there. You also have to subscribe your component to the `maxPopulation` store value, because otherwise, you can't make the comparison.

Now you'll likely abstract that check into a function or a custom hook that does the check:

```
const useConditionalIncreasePopulation = (by: number) => {
  const currentPopulation = useSelector(state => state.currentPopulation)
  const maxPopulation = useSelector(state => state.maxPopulation)
  const newPopulation = currentPopulation + maxPopulation

  return () => {
    store.setPopulation(Math.min(newPopulation, maxPopulation))
  }
}
```

and now we have a worse version of just using an event-based action in the first place 🤷

**Reply by lgenzelis** [reply]
This example was brilliant :)

**Reply by ssijak** [reply]
What is the percentage of state changes that fit that narrative and compared to where using a setter like function works? I would wager that much more cases fit setter way perfectly fine.

**Reply by TkDodo** [reply]
Most setter examples that I can come up with shouldn’t exist. For example, taking data from an API request and putting it into a store, with a function called `setTodos`. If you do that, yes, it’s a setter example, but imo you shouldn’t be doing that in the first place ;)

For things that are user interactions, which is what client stores are best used for, the event is literally what the user is doing. They aren’t “setting” a value, they are doing something with it. It could be as simple as “typing into an input field” to change a search string. The event is `search`, not `setSearch`. Applying a filter is `applyFilter`, not `setFilter` etc. Even if they just take a new value and put it into the store, they can evolve much better. And I don’t see why `setFilter(filter)` is better than `send({ type: 'applyFilter', filter })` 🤷‍♂️

**Reply by ssijak** [reply]
```And I don’t see why setFilter(filter) is better than send({ type: 'applyFilter', filter })```

Because I get triggered from my old Java days and having very complex configuration of Spring done in untyped magical XML files when it could have been just a set of normal function calls :) 
But yeah, mostly a preference.

**Reply by eug-vs** [reply]
This is how I see it: setter explicitly declares that it's *only* setting a value. Even if only the name of the function differs at the moment (as in example above `setFilter` / `applyFilter`), you immediately know that this function *might* do something else, not just set a value. You treat it like a blackbox and let it do it's thing. Your responsibility as a consumer in this case is just to tell the state machine what's going on on the outside (you send an event), and you can be sure it will respond accordingly. This prevents leaking your state machine logic to the outside and keeps your state machine logic verifiable independently from the consumer (you have finite number of events). This better applies to FSM, but I find this mental model suitable for other stores as well.

One analogy may be external API (maybe not the best one). Imagine if you have a 3rd party API that only gives you proxy interface to their database. This is hardly an API - now you have to write *their* SQL in *your* application code.

**Reply by gavinthomas-valtech** [reply]
This is the best response I have seen to this event driven API by David himself

https://github.com/statelyai/xstate/discussions/1755#discussioncomment-227702


>Let me explain one of the reasons I prefer service.send({ type: 'UPDATE_NAME' }) versus service.updateName(). Can you tell the difference between these two lines of code?

service.updateName();

service.updateName();

Do you see a difference? There might not be a difference at first sight, but these two method calls might have completely different behavior depending on when they are called. As a developer, when you see service.updateName(), you would expect, on the very next line, for the name to be updated. But, for example, that might not be the case if you are in a state where you are forbidden from updating the name. In that case, there's no way of knowing whether your updateName() imperative call succeeded or failed without checking the state.

On the other hand, service.send({ type: 'UPDATE_NAME' }) makes it very clear that you are sending an intent to update the name, and you know it 1) might not be immediate, and 2) might not succeed. That second part is very important - sending a message feels different than making an imperative method call, because all you're doing is sending a message. You are not directly manipulating an entity.

**Reply by gavinthomas-valtech** [reply]
>As a real-life analogy, consider these two different calls:

human.sleep();

human.send('GO_TO_SLEEP');

In the first one, you expect the method call to succeed; it feels like you're directly manipulating the human. But in the second one, you're merely telling the human to go to sleep. It doesn't feel direct, which is good - there is the understanding that sending a message doesn't mean that the desired outcome will occur, and that's okay.

**Comment by renet** [comment]
Thanks for this article. Would’ve loved to see a state machine example here, as well, because I’m one of those who shy away from that topic. But I assume a simple state machine example isn’t easily given.

I do have a question, which is more of a TypeScript question I suppose: Wouldn’t it be possible to use TypeScript in a way that from the transition definitions passed to the `createStore` function we create a store  object holding these transition functions as properties instead of having to call `store.send`. So for example `store.increasePopulation({ by: 10 })` (or similar) could be offered. This would be less Redux-ish so that Redux haters would be lacking at least one argument against xstate/store. I know this is not a feature, but could it be, with the power of TypeScript?

**Reply by TkDodo** [reply]
I think David is considering that, at least from what I've seen on Twitter

**Comment by eug-vs** [comment]
Thanks for the amazing article! I was just first trying `xstate` for my project recently (was not aware of `xstate/store` at that time).
You mentioned `react-hook-form`, would be great to see how `xstate` (or `xstate/store`) could play with more complex forms with highly conditional control flow. 
Real life example: a multi-step order cart form (client component since we want to display it in a mobile bottom drawer). Some steps are required, but some (like entering the delivery address) only make sense with certain order types. I had really rough time drawing the line between responsibilities of RHF and Xstate there, with questions like
 -  "should my state machine manage *only* presentation layer?" (i.e which form step UI should be displayed to the user)
 -  "should RHF submit a mini-form after each step and put it into a external client state?"
 -  "am I making a mistake by putting too much load onto a client side?"
 -  "should I consider RHF formState or Xstate context to be the source of truth?"

RHF has some say on such scenarios described here: https://react-hook-form.com/advanced-usage - they display an example with `little-state-machine`. I avoided this approach since I had already built some logic into my form schema (with some refines making sure that again e.g "delivery" order should have an address). Probably that was a mistake - but it had some benefits too (e.g server uses the same schema to immediately reject invalid orders).
What I'm sure about is that state machine was a *great choice* in such case. I have all the pieces, but putting them together *nicely* was a challenge. With respect to this article - will `xstate/store` help the matters? I'm so inspired to go refactor the code to use `xstate/store` right now, even though I've managed to avoid client side state managers up to this moment (apart from `react-query` ofc!).
Of course I understand it highly depends on the particular use-case and what the project is trying to achieve, but I would love to hear your thoughts on such cases.

P.S: Sorry for the huge dump! I think this just on the edge of on-topic for this article, and it was just soo well-timed I could not resist putting my thoughts here

**Reply by eug-vs** [reply]
Upon further exploration, I figured that my biggest concern was the choice between XState context VS RHF formState. I ended up declaring RHF formState as the source of truth and not duplicating any RHF form data to the XState context. Instead, I attach a snapshot of `form.getValues()` with each `event` sent to XState. Maybe it's a bit of an overkill, but this way I don't have a headache of syncing 2 states (xstate context and RHF). I still have access to all the data (I just pull it from event rather than context). In a sense this doesn't really differ that much, but it helps with the mental model a lot. Also it feels like having *less* context in XState makes it a bit more "pure".

This is not yet fully battle-tested, but this is what I ended up with, at least for now.

**Comment by thepradipvc** [comment]
What's the difference between store and state machines? And when do we actually need a state machine, can someone give an example?

**Reply by TkDodo** [reply]
For example, we have a sidebar that only had a collapse / expand button, so we started with some simple state. Needed that information available globally, so we moved to a store. Then, we added a 3rd state where the user could hover the collapsed sidebar to temporarily expand it, and UX wanted animations. Now it became more complex because you could click the expand button while you were hovering the sidebar etc. So it became a simple state machine with transitions and we didn't have to worry about edge cases where you would go from e.g. the `expanding` state to `collapsed` immediately because that wasn't allowed by the transitions.

**Comment by emadabdulrahim** [comment]
One of the things I like about Zustand is middleware. For example, I use the [persist](https://zustand.docs.pmnd.rs/integrations/persisting-store-data#persisting-store-data) middleware a lot. Is there any similar API for xstate/store?

**Reply by TkDodo** [reply]
I don't think so, at least not yet

**Reply by TkDodo** [reply]
David answered on [Twitter](https://x.com/DavidKPiano/status/1871176125071962466) that you can just do:

```
store.subscribe(state => {
  persist(state);
});
```

which is pretty great

**Comment by ruiaraujo012** [comment]
@TkDodo I think you should update the post as the API used to create a store is deprecaded: 
<img width="794" alt="image" src="https://github.com/user-attachments/assets/68a61fdc-9ef1-48cd-941a-593788f5fa4a">

**Reply by TkDodo** [reply]
yeah seems that it changed towards an `on` property. Pretty simple change - do you want to make a PR for that?

**Reply by ruiaraujo012** [reply]
Sure, I'll do that.

**Reply by ruiaraujo012** [reply]
Done. PR created.

**Comment by alexryanterry** [comment]
Are there any recommend patterns for using xstate with react-query or is the general idea that you use react-query to fetch data from your server and then feed that into the store?

For example, maybe there is a complicated form that requires xstate. The form data is fetched using react-query. When the form is submitted, it:

1. calls a mutation to update the data
2. invalidates the query and refreshes the data in xstate

Or maybe it doesn't even have to wait for a form submission.. Maybe it's a more dynamic form / editable data grid where mutations are being called on change / blur.

**Reply by TkDodo** [reply]
> or is the general idea that you use react-query to fetch data from your server and then feed that into the store?

no, the idea is never to synchronize data from react-query with another store, because that’s not necessary. See:

- https://tkdodo.eu/blog/react-query-as-a-state-manager
- https://tkdodo.eu/blog/thinking-in-react-query

for forms specifically, see:
- https://tkdodo.eu/blog/practical-react-query#keep-server-and-client-state-separate
- https://tkdodo.eu/blog/react-query-and-forms

**Reply by alexryanterry** [reply]
Okay, thanks for the quick reply! I'll have to spend some more time with these ideas. 

I keep hearing this from your blogs and everywhere people know what they're talking about so I know it is the way, but I have a hard time figuring out how to set up tightly integrated / complex interactions between UI state and server.

We're building fairly complex "forms", on the spectrum of excel-sheet editors, so most of the form examples I see don't get me quite there, but I'll keep at it.

**Comment by monarcode** [comment]
Is there a way to persist store to localstorage or something, using xtstae/store?

**Reply by TkDodo** [reply]
I don’t think so right now, no.

**Reply by TkDodo** [reply]
David answered on [Twitter](https://x.com/DavidKPiano/status/1871176125071962466) that you can just do:

```
store.subscribe(state => {
  persist(state);
});
```

which is pretty great

**Comment by herkulano** [comment]
How do you hydrate the store with the initial context? The docs have no mention of hydration

**Reply by alexryanterry** [reply]
UseEffect calls event to set context when the store is used in a component? Does xstate not document this pattern?

**Reply by herkulano** [reply]
useEffect() doesn't run on the server, so you are not setting the initial data, you're just sending an event to update the data after it has been initialized. 

For example if you have a date that is rendered on the server you need to set the time zone and formatting server side so that it's rendered on the server in the same way that is rendered on the client.

It's still not clear to me how you would achieve this with @xstate/store only.

**Reply by TkDodo** [reply]
you can pass an `initialContext` to `createStore`, both on the server and on the client. Can you show an example of what you’re having trouble with because I’m not fully understanding the problem

**Reply by herkulano** [reply]
I've created this StackBlitz: https://stackblitz.com/edit/remix-run-remix-rcseckqc?file=app%2Froot.tsx

In this Remix example I'm using `useEffect()` which gives a FOUC with the default currency and locale. If you refresh the page you'll see it, it's fast, if you blink you'll miss it, but it's really annoying.

What I'm still not understanding is how do I pass the `initialContext` to the `createStore()` with the data from the `loader` when it the `store` lives outside the React components?

For example, in Jotai I can use a `<Provider>` and hydrate the values with `useHydrateAtoms()`: https://jotai.org/docs/utilities/ssr

**Reply by TkDodo** [reply]
On the server, you probably need to get the locale by reading the request headers in the loader. Then, return them as `loaderData`, then create a context with the store on the client, similar to how I’d do that with zustand: https://tkdodo.eu/blog/zustand-and-react-context

**Reply by herkulano** [reply]
This is awesome! Thank you so much for being so helpful!

**Reply by herkulano** [reply]
The typing is a small annoyance though. Because the store is set inside the provider we need to type `useContext()` manually with the events.

```ts
type LocaleEvents = { type: "tz"; tz: string } | { type: "currency"; currency: string };
const LocaleStoreContext = createContext<Store<LocaleStore, LocaleEvents, EventObject> | null>(null);
```

**Comment by justice-sh** [comment]
Thanks for this article.

What do you think about having an option like: 

store.action.increasePopulation({ by: 10 })

So that people can choose to use either the event: store.send or the action. I do personally prefer the action.

**Reply by ColinMelendez** [reply]
Just wanted to point out that this does exist!

Stores have a fluent API as well as the declarative one, so these would be an equivalent way to trigger an event on a store:

```ts
// declarative api
store.send({ 
  type: “increasePopulation”,
  by: 10
});

// fluent API
store.trigger.increasePopulation({
  by: 10
});
```
